Tasks 1-10, 12-17 + 19

PutienkoEN opened the pull request:
Tasks 1-10 + 19
Tests pulled from master.

PutienkoEN committed:
Task 1 - 10
Task 19

Car class edited(changed private-package constructor to public)

PutienkoEN committed:
Added tests from master(for 1-10, 19)

nkarulin commented on src/main/java/com/epam/javacourses2016/task1/SolverTask1.java:
```diff
@@ -11,12 +16,47 @@
+        return lines;
+    }
+
+    private List<String> fileReader(File file) {
```
Not the best name for a method, `readFile` is better

nkarulin commented on src/main/java/com/epam/javacourses2016/task10/SolverTask10.java:
```diff
@@ -13,10 +17,35 @@
+            Pattern pattern = Pattern.compile("[(A-Za-z)('\\-)(A-Za-z)]+");
+            while ((line = reader.readLine()) != null) {
+                Matcher matcher = pattern.matcher(line);
+                while (matcher.find()) {
```
why don't we just use Scanner#next()?

PutienkoEN replied:
Yes, it will be better. Didn't notice Scanner#Next could use regex pattern for strings.
As for regex pattern, i thought we should remove numbers from text if there is some(1,2,3 etc).

nkarulin commented on src/main/java/com/epam/javacourses2016/task19/SolverTask19.java:
```diff
@@ -17,8 +17,36 @@
+        int count = 0;
+        for (Car car1 : cars) {
+            for (Car car2 : cars) {
+                count += getOvertake(car1, car2, lengthLap, numberLaps);
```
too slow) :snail: 
You simulate a race for each pair of cars instead of the single race simulation for everyone

nkarulin commented on src/main/java/com/epam/javacourses2016/task5/SolverTask5.java:
```diff
@@ -1,24 +1,63 @@
+            eBig = eBig.add(funcBig.pow(2));
+        }
+
+        return eBig.doubleValue();
```
nice :+1:

nkarulin commented on src/main/java/com/epam/javacourses2016/task8/SolverTask8.java:
```diff
@@ -10,12 +12,59 @@
+        Stack<Character> characters = new Stack<>();
+
+        char[] openTypes = new char[]{'(', '[', '{'};
+        char[] closingTypes = new char[]{')', ']', '}'};
```
using Sets instead of arrays will make the code shorter and cleaner

PutienkoEN replied:
Not shure how i can do this with set, i need to know index of open braket and index of closed braket. Or i should iterate through both of them to calculate it like i'm doing it with arrays?

nkarulin replied:
You can just ask set: "Have you got such a bracket?", no iteration is needed

PutienkoEN replied:
There could be situation like {some text] - and if i ask set .contains() it give me true, but it will be wrong answer. I have to check pairs also {} [] etc, thats why i need to know index. I thought about map, but i feel like it will be slower than array of chars.

nkarulin replied:
Ok then
But you are wrong about map is slower. In-array search is linear, search in hash map is constant (if no collisions)

nkarulin commented on src/main/java/com/epam/javacourses2016/task9/SolverTask9.java:
```diff
@@ -13,10 +17,25 @@
+        try (BufferedReader reader = new BufferedReader(new FileReader(input))) {
+            String line;
+            Pattern pattern = Pattern.compile("[(A-Za-z)('\\-)(A-Za-z)]+");
+            while ((line = reader.readLine()) != null) {
```
why dont we just use Scanner#next()?

PutienkoEN committed:
Added:
Task 12, 13, 14, 15, 16, 17

Edited:
Task 2(removed static from method)
Task 5(edited formula)
Task 9, 10 - added Scanner (simplified code)
Task 19 - edited algorithm (was wrong formula).

Tests added:
11, 14, 15, 17

Point2d, Segment - added equals and hascode methods.

PutienkoEN committed:
Merge branch 'master-2' into PE

PutienkoEN commented:
Added: Task 12, 13, 14, 15, 16, 17

Edited:
Task 2(removed static from method)
Task 5(edited formula)
Task 9, 10 - added Scanner (simplified code)
Task 19 - edited algorithm (was wrong formula).

Tests added:
11, 14, 15, 17

Point2d, Segment - added equals and hascode methods.

nkarulin commented on README.md:
```diff
@@ -1,84 +0,0 @@
-
-## Баги, предложения, вопросы
-Подобные дела следует оформлять в виде Issue в главном репозитории. Так же не стесняйтесь задавать возникающие вопросы по флоу\гиту мне в почту. 
-
```
Dude, why do you delete my README? :cry:

PutienkoEN replied:
Ups.. It will be back soon!

nkarulin commented on src/main/java/com/epam/javacourses2016/task12/SolverTask12.java:
```diff
@@ -1,22 +1,40 @@
+
+        int indexStart = integers.size() - 1;
+
+        for (int i = 0; i < integers.size(); i++) {
```
Use iterator instead, with LinkingList such loop will give us O(logn)

PutienkoEN replied:
Ok. I got it. And what about new Array List with initial capacity, it will be better than LinkedList? 
Smth like this:
```
        List<Integer> arrayList = new ArrayList<>(integers.size());
        int beginIndex = 0;
        int endIndex = integers.size() - 1;

        for (int i = 0; i < integers.size(); i++) {
            int listValue = integers.get(i);
            if (listValue > value) {
                arrayList.add(endIndex, listValue);
                endIndex--;
            } else {
                arrayList.add(beginIndex, listValue);
                beginIndex++;
            }
        }
```

nkarulin commented on src/main/java/com/epam/javacourses2016/task12/SolverTask12.java:
```diff
@@ -1,22 +1,40 @@
+
+        //Collection sort could solve this problem, so here we go!
+        //Collections.sort(integers);
+        return integers;
```
You'll better use custom comparator in that task

nkarulin commented on src/main/java/com/epam/javacourses2016/task13/MyGraphCreator.java:
```diff
@@ -0,0 +1,116 @@
+        public void addEdge(int first, int second) {
+
+            if (!(checkInputValue(first) && checkInputValue(second))) {
+                return;
```
It seems like exception situation

nkarulin commented on src/main/java/com/epam/javacourses2016/task13/MyGraphCreator.java:
```diff
@@ -0,0 +1,116 @@
+            }
+
+            if (nodes.size() >= NUMBER_NODES) {
+                return;
```
And it seems like exception situation

nkarulin commented on src/main/java/com/epam/javacourses2016/task13/MyGraphCreator.java:
```diff
@@ -0,0 +1,116 @@
+        @Override
+        public void removeEdge(int first, int second) {
+            if (!(checkInputValue(first) && checkInputValue(second))) {
+                return;
```
And so on

nkarulin commented on src/main/java/com/epam/javacourses2016/task13/MyGraphCreator.java:
```diff
@@ -0,0 +1,116 @@
+                    secondNode.getConnectedNodes().contains(firstNode);
+        }
+
+        private boolean checkInputValue(int value) {
```
it can be static

PutienkoEN replied:
But there is NUMBER_NODES call, and it's not static.

nkarulin replied:
Right. But if `NUMBER_NODES` is not static (constant), why it's written in capital letters?

nkarulin commented on src/main/java/com/epam/javacourses2016/task13/MyGraphCreator.java:
```diff
@@ -0,0 +1,116 @@
+            return true;
+        }
+
+        class Node {
```
Can not it be static?

PutienkoEN replied:
Node class? Idea telling me "inner class can not have static declaration". I'l read about it more, not often use nested/inner classes.

nkarulin replied:
We had lecture about inner/nested classes.
So, if Node does not use anything from MyGraphCreator object it should not keep link to it, right?

nkarulin commented on src/main/java/com/epam/javacourses2016/task14/MyCollection.java:
```diff
@@ -0,0 +1,102 @@
+/**
+ * Created by Elarion on 25.11.2016.
+ */
+public class MyCollection<T extends Number> implements AbstractCollectionCreator.NumberCollection<T> {
```
You could use AbstractCollection as parent class, less overrides would happens

nkarulin commented on src/main/java/com/epam/javacourses2016/task17/SolverTask17.java:
```diff
@@ -18,8 +21,76 @@
+        Set<Point2D> resultSet = new HashSet<>();
+        double minX = Double.MAX_VALUE;
+
+        for (int i = 0; i < segments.size(); i++) {
```
:snail: What have i said about sorting first?

PutienkoEN replied:
I added sort now. Also added some breaks from iteration. Also thought about TreeSet here, because i dont rly need map value here(or i missing somethin).

PutienkoEN committed:
Task 1 renamed method(fileReader->readFile
Task 12 - linked list with iterator.
GraphCreator - exceptions added
MyCollection - extends AbstractCollection.

Readme back!

PutienkoEN committed:
added addAll to MyCollection

PutienkoEN committed:
Edited Task 17 - Added sort + other stuff.

bmkodoo commented:
Master was updated. Resolve conflicts please.
Please stop contribute to the branch till pull request will be merged, it become huge;)

Nice work, ready to merge but surprised to see tests 11, 14 and 15 here

PutienkoEN commented:
So i need only resolve conflicts, and wait till you merge it? And later i will fix some mistakes with new PR?
Or i can fix it now, but it will add more to this pr... i'm confused.
About tests - I wanted push tests to master first, but was not shure about this tests. 11 task not done yet, not shure how i should test it, 14 creating my extention of AbstractСollection => for other it will not work because everyone call their classes different. Not shure what 15 doing here, my miss.

PutienkoEN committed:
Task 17, Task 5Test solved master conflicts

nkarulin commented on src/main/java/com/epam/javacourses2016/task17/SolverTask17.java:
```diff
@@ -18,8 +21,154 @@
+        double minX = Double.MAX_VALUE;
+
+        mainLoop:
+        for (Segment segmentOne : segmentsMap.keySet()) {
```
:slow: Be engineer, not coder!
[Пересечение множества отрезков](https://neerc.ifmo.ru/wiki/index.php?title=%D0%9F%D0%B5%D1%80%D0%B5%D1%81%D0%B5%D1%87%D0%B5%D0%BD%D0%B8%D0%B5_%D0%BC%D0%BD%D0%BE%D0%B6%D0%B5%D1%81%D1%82%D0%B2%D0%B0_%D0%BE%D1%82%D1%80%D0%B5%D0%B7%D0%BA%D0%BE%D0%B2), [Алгоритм Бентли-Оттмана](https://neerc.ifmo.ru/wiki/index.php?title=%D0%90%D0%BB%D0%B3%D0%BE%D1%80%D0%B8%D1%82%D0%BC_%D0%91%D0%B5%D0%BD%D1%82%D0%BB%D0%B8-%D0%9E%D1%82%D1%82%D0%BC%D0%B0%D0%BD%D0%B0)
